File: src/main/java/essentials/core/plugin/PluginData.java
package essentials.core.plugin;

import arc.struct.Array;
import arc.util.serialization.Json;
import essentials.internal.Log;
import mindustry.world.Tile;
import mindustry.world.blocks.logic.MessageBlock;
import org.hjson.JsonArray;
import org.hjson.JsonObject;
import org.hjson.JsonValue;

import java.io.*;
import java.time.LocalDateTime;

import static essentials.Main.root;
import static mindustry.Vars.world;

public class PluginData {
    private final Json json = new Json();

    // 일회성 플러그인 데이터
    public Array<nukeblock> nukeblock = new Array<>();
    public Array<eventservers> eventservers = new Array<>();
    public Array<powerblock> powerblock = new Array<>();
    public Array<messagemonitor> messagemonitor = new Array<>();
    public Array<messagewarp> messagewarp = new Array<>();
    public Array<Tile> scancore = new Array<>();
    public Array<Tile> nukedata = new Array<>();
    public Array<Tile> nukeposition = new Array<>();
    public Array<Process> process = new Array<>();

    // 종료시 저장되는 플러그인 데이터
    public Array<warpzone> warpzones = new Array<>();
    public Array<warpblock> warpblocks = new Array<>();
    public Array<warpcount> warpcounts = new Array<>();
    public Array<warptotal> warptotals = new Array<>();
    public Array<String> blacklist = new Array<>();
    public Array<banned> banned = new Array<>();

    public void saveAll() {
        JsonObject data = new JsonObject();
        JsonArray buffer = new JsonArray();
        for (int a = 0; a < warpzones.size; a++) {
            buffer.add(JsonValue.readHjson(json.prettyPrint(warpzones.get(a))));
        }
        data.add("warpzones", buffer);
        buffer = new JsonArray();
        for (int a = 0; a < warpblocks.size; a++) {
            buffer.add(JsonValue.readHjson(json.prettyPrint(warpblocks.get(a))));
        }
        data.add("warpblocks", buffer);
        buffer = new JsonArray();
        for (int a = 0; a < warpcounts.size; a++) {
            buffer.add(JsonValue.readHjson(json.prettyPrint(warpcounts.get(a))));
        }
        data.add("warpcounts", buffer);
        buffer = new JsonArray();
        for (int a = 0; a < warptotals.size; a++) {
            buffer.add(JsonValue.readHjson(json.prettyPrint(warptotals.get(a))));
        }
        data.add("warptotals", buffer);
        buffer = new JsonArray();
        for (int a = 0; a < blacklist.size; a++) {
            buffer.add(blacklist.get(a));
        }
        data.add("blacklist", buffer);
        buffer = new JsonArray();
        for (int a = 0; a < banned.size; a++) {
            buffer.add(JsonValue.readHjson(json.prettyPrint(banned.get(a))));
        }
        data.add("banned", buffer);
        root.child("data/PluginData.object").writeString(data.toString(), false);
    }

    public void loadall() {
        try {
            if (!root.child("data/PluginData.object").exists()) {
                saveAll();
            } else {
                JsonObject data = JsonValue.readJSON(root.child("data/PluginData.object").readString()).asObject();
                for (int a = 0; a < data.get("warpzones").asArray().size(); a++) {
                    JsonObject buffer = data.get("warpzones").asArray().get(a).asObject();
                    warpzones.add(new warpzone(
                            buffer.get("mapName").asString(),
                            buffer.get("startx").asInt(),
                            buffer.get("starty").asInt(),
                            buffer.get("finishx").asInt(),
                            buffer.get("finishy").asInt(),
                            buffer.get("touch").asBoolean(),
                            buffer.get("ip").asString(),
                            buffer.get("port").asInt()
                    ));
                }

                for (int a = 0; a < data.get("warpblocks").asArray().size(); a++) {
                    JsonObject buffer = data.get("warpblocks").asArray().get(a).asObject();
                    warpblocks.add(new warpblock(
                            buffer.get("mapName").asString(),
                            buffer.get("tilex").asInt(),
                            buffer.get("tiley").asInt(),
                            buffer.get("tileName").asString(),
                            buffer.get("size").asInt(),
                            buffer.get("ip").asString(),
                            buffer.get("port").asInt(),
                            buffer.get("description").asString()
                    ));
                }

                for (int a = 0; a < data.get("warpcounts").asArray().size(); a++) {
                    JsonObject buffer = data.get("warpcounts").asArray().get(a).asObject();
                    warpcounts.add(new warpcount(
                            buffer.get("mapName").asString(),
                            buffer.get("x").asInt(),
                            buffer.get("y").asInt(),
                            buffer.get("ip").asString(),
                            buffer.get("port").asInt(),
                            buffer.get("players").asInt(),
                            buffer.get("numbersize").asInt()
                    ));
                }

                for (int a = 0; a < data.get("warptotals").asArray().size(); a++) {
                    JsonObject buffer = data.get("warptotals").asArray().get(a).asObject();
                    warptotals.add(new warptotal(
                            buffer.get("mapName").asString(),
                            buffer.get("x").asInt(),
                            buffer.get("y").asInt(),
                            buffer.get("totalplayers").asInt(),
                            buffer.get("numbersize").asInt()
                    ));
                }

                for (int a = 0; a < data.get("blacklist").asArray().size(); a++) {
                    JsonArray buffer = data.get("warpzones").asArray().get(a).asArray();
                    blacklist.add(buffer.get(a).asString());
                }

                for (int a = 0; a < data.get("banned").asArray().size(); a++) {
                    JsonObject buffer = data.get("banned").asArray().get(a).asObject();
                    banned.add(new banned(
                            buffer.get("time").asString(),
                            buffer.get("name").asString(),
                            buffer.get("uuid").asString(),
                            buffer.get("reason").asString()
                    ));
                }
                Log.info("plugindata-loaded");
            }
        } catch (Exception i) {
            i.printStackTrace();
            root.child("data/PluginData.object").delete();
        }
    }

    private static byte[] serialize(Object obj) throws IOException {
        ByteArrayOutputStream b = new ByteArrayOutputStream();
        ObjectOutputStream o = new ObjectOutputStream(b);
        o.writeObject(obj);
        return b.toByteArray();
    }

    public static Object deserialize(byte[] bytes) throws IOException, ClassNotFoundException {
        ByteArrayInputStream b = new ByteArrayInputStream(bytes);
        ObjectInputStream o = new ObjectInputStream(b);
        return o.readObject();
    }

    public static class nukeblock {
        public final Tile tile;
        public final String name;

        public nukeblock(Tile tile, String name) {
            this.tile = tile;
            this.name = name;
        }
    }

    public static class eventservers {
        public final String roomname;
        public int port;

        public eventservers(String roomname, int port) {
            this.roomname = roomname;
            this.port = port;
        }
    }

    public static class powerblock {
        public final Tile messageblock;
        public final Tile tile;

        public powerblock(Tile messageblock, Tile tile) {
            this.messageblock = messageblock;
            this.tile = tile;
        }
    }

    public static class messagemonitor {
        public final MessageBlock.MessageBlockEntity entity;

        public messagemonitor(MessageBlock.MessageBlockEntity entity) {
            this.entity = entity;
        }
    }

    public static class messagewarp {
        public final Tile tile;
        public final String message;

        public messagewarp(Tile tile, String message) {
            this.tile = tile;
            this.message = message;
        }
    }

    public static class warpzone {
        public final String mapName;
        public final int startx;
        public final int starty;
        public final int finishx;
        public final int finishy;
        public final String ip;
        public final int port;
        public final boolean touch;

        public warpzone(String mapName, Tile start, Tile finish, boolean touch, String ip, int port) {
            this.mapName = mapName;
            this.startx = start.x;
            this.starty = start.y;
            this.finishx = finish.x;
            this.finishy = finish.y;
            this.ip = ip;
            this.port = port;
            this.touch = touch;
        }

        private warpzone(String mapName, int startx, int starty, int finishx, int finishy, boolean touch, String ip, int port) {
            this.mapName = mapName;
            this.startx = startx;
            this.starty = starty;
            this.finishx = finishx;
            this.finishy = finishy;
            this.ip = ip;
            this.port = port;
            this.touch = touch;
        }

        public Tile getStartTile() {
            return world.tile(startx, starty);
        }

        public Tile getFinishTile() {
            return world.tile(finishx, finishy);
        }
    }

    public static class warpblock {
        public final String mapName;
        public final int tilex;
        public final int tiley;
        public final String tileName;
        public final String ip;
        public final int port;
        public final String description;
        public final int size;

        public warpblock(String mapName, Tile tile, String ip, int port, String description) {
            this.mapName = mapName;

            this.ip = ip;
            this.port = port;
            this.description = description;

            this.tilex = tile.entity.tileX();
            this.tiley = tile.entity.tileY();
            this.tileName = tile.entity.block.name;
            this.size = tile.entity.block.size;
        }

        private warpblock(String mapName, int tilex, int tiley, String tileName, int size, String ip, int port, String description) {
            this.mapName = mapName;

            this.ip = ip;
            this.port = port;
            this.description = description;

            this.tilex = tilex;
            this.tiley = tiley;
            this.tileName = tileName;
            this.size = size;
        }
    }

    public static class warpcount {
        public final String mapName;
        public final int x;
        public final int y;
        public final String ip;
        public final int port;
        public int players;
        public int numbersize;

        public warpcount(String mapName, Tile tile, String ip, int port, int players, int numbersize) {
            this.mapName = mapName;
            this.x = tile.x;
            this.y = tile.y;
            this.ip = ip;
            this.port = port;
            this.players = players;
            this.numbersize = numbersize;
        }

        private warpcount(String mapName, int x, int y, String ip, int port, int players, int numbersize) {
            this.mapName = mapName;
            this.x = x;
            this.y = y;
            this.ip = ip;
            this.port = port;
            this.players = players;
            this.numbersize = numbersize;
        }

        public Tile getTile() {
            return world.tile(x, y);
        }
    }

    public static class warptotal {
        public final String mapName;
        public final int x;
        public final int y;
        public int totalplayers;
        public int numbersize;

        public warptotal(String mapName, Tile tile, int totalplayers, int numbersize) {
            this.mapName = mapName;
            this.x = tile.x;
            this.y = tile.y;
            this.totalplayers = totalplayers;
            this.numbersize = numbersize;
        }

        private warptotal(String mapName, int x, int y, int totalplayers, int numbersize) {
            this.mapName = mapName;
            this.x = x;
            this.y = y;
            this.totalplayers = totalplayers;
            this.numbersize = numbersize;
        }

        public Tile getTile() {
            return world.tile(x, y);
        }
    }

    public static class banned {
        public final String time;
        public final String name;
        public final String uuid;
        public final String reason;

        public banned(LocalDateTime time, String name, String uuid, String reason) {
            this.time = time.toString();
            this.name = name;
            this.uuid = uuid;
            this.reason = reason;
        }

        private banned(String time, String name, String uuid, String reason) {
            this.time = time;
            this.name = name;
            this.uuid = uuid;
            this.reason = reason;
        }

        public LocalDateTime getTime() {
            return LocalDateTime.parse(time);
        }
    }
}

File: src/main/kotlin/essentials/PluginData.kt
package essentials

import arc.struct.ObjectMap
import arc.struct.Seq
import arc.util.Log
import mindustry.Vars
import mindustry.world.Tile
import org.hjson.JsonArray
import org.hjson.JsonObject
import org.jetbrains.exposed.sql.insert
import org.jetbrains.exposed.sql.selectAll
import org.jetbrains.exposed.sql.transactions.transaction
import java.io.IOException
import java.util.*
import java.util.regex.Pattern

object PluginData {
    var uptime = 0L
    var playtime = 0L
    var pluginVersion = ""

    var warpZones = Seq<WarpZone>()
    var warpBlocks = Seq<WarpBlock>()
    var warpCounts = Seq<WarpCount>()
    var warpTotals = Seq<WarpTotal>()
    var blacklist = Seq<Pattern>()
    var banned = Seq<Banned>()
    var status = ObjectMap<String, String>()

    var sudoPassword = ""

    var lastMemory = ""
    var vpnList = Seq<String>()
    var changed = false
    var isRankingWorking = false

    data class WarpZone(val mapName : String, val start : Int, val finish : Int, val touch : Boolean, val ip : String, val port : Int) {
        val startTile : Tile get() = Vars.world.tile(start)
        val finishTile : Tile get() = Vars.world.tile(finish)
    }

    data class WarpBlock(val mapName : String, val x : Int, val y : Int, val tileName : String, val size : Int, val ip : String, val port : Int, val description : String) {
        var online = false
    }

    data class WarpCount(val mapName : String, val pos : Int, val ip : String, val port : Int, var players : Int, var numbersize : Int) {
        val tile : Tile get() = Vars.world.tile(pos)
    }

    data class WarpTotal(val mapName : String, val pos : Int, var totalplayers : Int, var numbersize : Int) {
        val tile : Tile get() = Vars.world.tile(pos)
    }

    data class Banned(val time : Long, val name : String, val uuid : String, val reason : String)

    operator fun get(key : String) : String? {
        return status.get(key)
    }

    fun save(first: Boolean) {
        val data = JsonObject()
        var buffer = JsonArray()

        for(it in warpZones) {
            val obj = JsonObject()
            obj.add("mapName", it.mapName)
            obj.add("start", it.start)
            obj.add("finish", it.finish)
            obj.add("touch", it.touch)
            obj.add("ip", it.ip)
            obj.add("port", it.port)
            buffer.add(obj)
        }
        data.add("warpZones", buffer)
        buffer = JsonArray()

        for(it in warpBlocks) {
            val obj = JsonObject()
            obj.add("mapName", it.mapName)
            obj.add("x", it.x)
            obj.add("y", it.y)
            obj.add("tileName", it.tileName)
            obj.add("size", it.size)
            obj.add("ip", it.ip)
            obj.add("port", it.port)
            obj.add("description", it.description)
            buffer.add(obj)
        }
        data.add("warpBlocks", buffer)
        buffer = JsonArray()

        for(it in warpCounts) {
            val obj = JsonObject()
            obj.add("mapName", it.mapName)
            obj.add("pos", it.pos)
            obj.add("ip", it.ip)
            obj.add("port", it.port)
            obj.add("players", it.players)
            obj.add("numbersize", it.numbersize)
            buffer.add(obj)
        }
        data.add("warpCounts", buffer)
        buffer = JsonArray()

        for(it in warpTotals) {
            val obj = JsonObject()
            obj.add("mapName", it.mapName)
            obj.add("pos", it.pos)
            obj.add("totalplayers", it.totalplayers)
            obj.add("numbersize", it.numbersize)
            buffer.add(obj)
        }
        data.add("warpTotals", buffer)
        buffer = JsonArray()

        for(it in blacklist) buffer.add(it.pattern())
        data.add("blacklist", buffer)
        buffer = JsonArray()

        for(it in banned) {
            val obj = JsonObject()
            obj.add("time", it.time)
            obj.add("name", it.name)
            obj.add("uuid", it.uuid)
            obj.add("reason", it.reason)
            buffer.add(obj)
        }
        data.add("banned", buffer)

        val json = JsonObject()
        for(it in status) {
            json.add(it.key, it.value)
        }
        data.add("status", json.toString())

        val encode = Base64.getEncoder()
        lastMemory = encode.encodeToString(data.toString().toByteArray())

        if (first) {
            transaction {
                DB.Data.insert {
                    it[DB.Data.data] = lastMemory
                }
            }
        }
    }

    fun load() {
        try {
            if(transaction { DB.Data.selectAll().empty() }) {
                save(true)
            } else {
                warpZones = Seq<WarpZone>()
                warpBlocks = Seq<WarpBlock>()
                warpCounts = Seq<WarpCount>()
                warpTotals = Seq<WarpTotal>()
                blacklist = Seq<Pattern>()
                banned = Seq<Banned>()
                status = ObjectMap<String, String>()

                transaction {
                    if(!DB.Data.selectAll().empty()) {
                        DB.Data.selectAll().first().run {
                            val data = JsonObject.readJSON(String(Base64.getDecoder().decode(this[DB.Data.data]))).asObject()

                            data["warpZones"].asArray().forEach {
                                val obj = it.asObject()
                                warpZones.add(WarpZone(obj.get("mapName").asString(), obj.get("start").asInt(), obj.get("finish").asInt(), obj.get("touch").asBoolean(), obj.get("ip").asString(), obj.get("port").asInt()))
                            }

                            data["warpBlocks"].asArray().forEach {
                                val obj = it.asObject()
                                warpBlocks.add(WarpBlock(obj.get("mapName").asString(), obj.get("x").asInt(), obj.get("y").asInt(), obj.get("tileName").asString(), obj.get("size").asInt(), obj.get("ip").asString(), obj.get("port").asInt(), obj.get("description").asString()))
                            }

                            data["warpCounts"].asArray().forEach {
                                val obj = it.asObject()
                                warpCounts.add(WarpCount(obj.get("mapName").asString(), obj.get("pos").asInt(), obj.get("ip").asString(), obj.get("port").asInt(), obj.get("players").asInt(), obj.get("numbersize").asInt()))
                            }

                            data["warpTotals"].asArray().forEach {
                                val obj = it.asObject()
                                warpTotals.add(WarpTotal(obj.get("mapName").asString(), obj.get("pos").asInt(), obj.get("totalplayers").asInt(), obj.get("numbersize").asInt()))
                            }

                            data["blacklist"].asArray().forEach { blacklist.add(Pattern.compile(it.asString())) }

                            data["banned"].asArray().forEach {
                                val obj = it.asObject()
                                banned.add(Banned(obj.get("time").asLong(), obj.get("name").asString(), obj.get("uuid").asString(), obj.get("reason").asString()))
                            }

                            try {
                                JsonArray.readJSON(data["status"].asString().replace("\\", "")).asObject().forEach {
                                    status.put(it.name, it.value.asString())
                                }
                            } catch(e : Exception) {
                                data["status"].asArray().forEach {
                                    if(!it.asString().equals("hubMode")) status.put(it.asString(), "none")
                                    if(it.asString().equals("hubMode")) Log.warn(Bundle()["event.plugin.hubmode.reset"])
                                }
                                changed = true
                            }
                        }
                    } else if(DB.Data.selectAll().empty()) {
                        DB.Data.insert {
                            it[data] = lastMemory
                        }
                    }
                }
            }
        } catch(e : IOException) {
            e.printStackTrace()
        } catch(e : Exception) {
            e.printStackTrace()
        }
    }
}